### PR TITLE
PIMOB-1266: Update header view on scroll

### PR DESCRIPTION
**Pull request by ehab-al-cko**
[Jira ticket](https://checkout.atlassian.net/browse/PIMOB-1266?atlOrigin=eyJpIjoiZGZhZTU3NjZmNWRmNDJkY2FkMzc3ZDBkNDJhYzYyMjkiLCJwIjoiaiJ9)
## Proposed changes

- Update header view background color
- Show title on navigation bar on scroll 
- Update Payment screen background color
- Add all card scheme icons into Custom 1 
- Remove subtitle from Custom1
- Remove all card scheme  from Custom2

https://user-images.githubusercontent.com/102961713/182137920-1d1c9cd4-b26b-4e2a-ad18-c396f833fba9.mov

## Types of changes

What types of changes does your code introduce to Frames Ios?
_Put an `x` in the boxes that apply_

* [ ] Bugfix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

* [ ] Lint and unit tests pass locally with my changes
* [ ] I have added tests that prove my fix is effective or that my feature works
* [ ] I have added necessary documentation (if appropriate)
* [ 
## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: Source/UI/NewUI/PaymentForm/View/PaymentHeaderView.swift [hunk 1]
```diff
@@ -11,7 +11,11 @@ import UIKit
 public final class PaymentHeaderView: UIView {
   private var style: PaymentHeaderCellStyle?
 
-  private var stackView: UIStackView = {
+  private lazy var backgroundView: UIView = {
+    UIView().disabledAutoresizingIntoConstraints()
+  }()
```
Comment (alex-ioja-yang-cko) [inline Source/UI/NewUI/PaymentForm/View/PaymentHeaderView.swift:16]: Do we need this "background view"? Looks like a complicated way to let the owner of the view to set it up as it wants.

Have you tried to just add paddings around the headerView inside PaymentViewController?

File: Source/UI/NewUI/PaymentForm/ViewController/PaymentViewController.swift [hunk 2]
```diff
@@ -148,6 +154,8 @@ final class PaymentViewController: UIViewController {
 extension PaymentViewController {
   private func setupViewModel() {
     delegate = self.viewModel as? PaymentViewControllerDelegate
+    view.backgroundColor = viewModel.paymentFormStyle?.backgroundColor
+    stackView.backgroundColor = viewModel.paymentFormStyle?.backgroundColor
```
Comment (alex-ioja-yang-cko) [inline Source/UI/NewUI/PaymentForm/ViewController/PaymentViewController.swift:158]: this method is called `setupViewModel`, but we're setting up the view instead. Can we find a more logical location for this code?

File: Source/UI/NewUI/PaymentForm/ViewModel/PaymentViewModel.swift [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import Checkout
+import UIKit
```
Comment (alex-ioja-yang-cko) [inline Source/UI/NewUI/PaymentForm/ViewModel/PaymentViewModel.swift:2]: ?

File: iOS Example Frame/iOS Example Frame/CustomStyle2.swift [hunk 4]
```diff
@@ -41,10 +42,7 @@ struct PaymentHeaderCellStyleCustom2: PaymentHeaderCellStyle {
   var backgroundColor = Constants.greenBackgroundColor
   var headerLabel: ElementStyle? = HeaderLabelFormStyleCustom2(text: "Payment details")
   public var subtitleLabel: ElementStyle? = TitleLabelStyleCustom2(text: "Visa, Mastercard and American Express accepted.")
-  public var schemeIcons: [UIImage?]? = [
-    UIImage(named: "visa-icon"),
-    UIImage(named: "mastercard-icon"),
-    UIImage(named: "amex-icon")]
+  public var schemeIcons: [UIImage?]? = nil
```
Comment (alex-ioja-yang-cko) [inline iOS Example Frame/iOS Example Frame/CustomStyle2.swift:45]: Do we have some task somewhere to fix this optional array of optional images? We really should be able to reduce at least one of the optionals, if not even both 🤔

File: Source/UI/NewUI/PaymentForm/View/PaymentHeaderView.swift (hunk 1 above)
Comment (ehab-al-cko) [reply Source/UI/NewUI/PaymentForm/View/PaymentHeaderView.swift:16]: Yes, we need it for header view.

File: iOS Example Frame/iOS Example Frame/CustomStyle2.swift (hunk 4 above)
Comment (ehab-al-cko) [reply iOS Example Frame/iOS Example Frame/CustomStyle2.swift:45]: There're 2 cases: 
1- The merchant may not want to show any images then  they can set it to `= nil`
2- `UIImage` is optional by default. they may send an image which was removed by mistake from their side then we don't show it.

Comment (alex-ioja-yang-cko) [reply iOS Example Frame/iOS Example Frame/CustomStyle2.swift:45]: 1. Wouldn't they just provide 0 images (ie: `[]`) ? Swift encourages removing optionality where possible so I would think that an empty array will always be easier to deal with

File: Source/UI/NewUI/PaymentForm/ViewController/PaymentViewController.swift [hunk 5]
```diff
@@ -103,7 +104,12 @@ final class PaymentViewController: UIViewController {
   }
 
   private func setupNavigationBar() {
-    customizeNavigationBarAppearance(color: viewModel.paymentFormStyle?.headerView.backgroundColor ?? .white)
+    let backgroundColor = viewModel.paymentFormStyle?.headerView.backgroundColor ?? .white
+    let titleColor = viewModel.paymentFormStyle?.headerView.headerLabel?.textColor ?? .black
+
+    let titleFont = viewModel.paymentFormStyle?.headerView.headerLabel?.font.withSize(15) ?? UIFont.systemFont(ofSize: 15)
```
Comment (alex-ioja-yang-cko) [inline Source/UI/NewUI/PaymentForm/ViewController/PaymentViewController.swift:110]: Should we hardcode the size? May allow consumer to set it themselves and own size?

```suggestion
    let titleFont = viewModel.paymentFormStyle?.headerView.headerLabel?.font ?? UIFont.systemFont(ofSize: Constants.defaultNavigationHeaderFontSize)
```

**Review by alex-ioja-yang-cko — APPROVED** [review]
Thanks for the explanations on the hierarchy earlier today 👍

File: Source/UI/NewUI/PaymentForm/View/PaymentHeaderView.swift [hunk 6]
```diff
@@ -86,11 +90,25 @@ public final class PaymentHeaderView: UIView {
...
     setupMainStackView()
     setupIconStackView()
     addArrangedSubview()
   }
 
+  private func setupBackgroundView() {
+    addSubview(backgroundView)
+    NSLayoutConstraint.activate([
+      backgroundView.topAnchor.constraint(equalTo: topAnchor,
+                                          constant: -Constants.Padding.l.rawValue),
```
Comment (deepesh-vasthimal-cko) [inline Source/UI/NewUI/PaymentForm/View/PaymentHeaderView.swift:103]: What is `l` ? in Constants.Padding.l. Can you rename it to having a longer name for readability ?

File: Source/UI/NewUI/PaymentForm/ViewController/PaymentViewController.swift [hunk 7]
```diff
@@ -297,3 +314,21 @@ extension PaymentViewController: SecurityCodeViewDelegate {
...
+
+extension PaymentViewController: UIScrollViewDelegate {
+  func scrollViewDidScroll(_ scrollView: UIScrollView) {
+    var contentOffsetY = scrollView.contentOffset.y
+
+    if #available(iOS 11.0, *) {
+      contentOffsetY += scrollView.adjustedContentInset.top
+    }
+
+    if headerView.frame.maxY > 0, contentOffsetY > headerView.frame.maxY / 2 {
```
Comment (deepesh-vasthimal-cko) [inline Source/UI/NewUI/PaymentForm/ViewController/PaymentViewController.swift:326]: Can we somehow unit test this logic ? what happens when scrollViewDidScroll

File: Source/UI/NewUI/PaymentForm/View/PaymentHeaderView.swift (hunk 6 above)
Comment (alex-ioja-yang-cko) [reply Source/UI/NewUI/PaymentForm/View/PaymentHeaderView.swift:103]: This has already been used extensively in the project so could be a larger refactor if done. Wouldn't want that to happen here as we cannot measure.

The Padding enum uses abbreviated common form variables to control number of paddings:

```swift
    public enum Padding: Double {

        /// parameter size: 2
        case xxs = 2

        /// parameter size: 8
        case xs = 8

        /// parameter size: 10
        case s = 10

        /// parameter size: 16
        case m = 16

        /// parameter size: 20
        case l = 20

        /// parameter size: 24
        case xl = 24

        /// parameter size: 28
        case xxl = 28

        /// parameter size: 40
        case xxxl = 40
    }
```

Might need to discuss and plan a route if we think an improvement is necessary, but for this PR this should not be relevant :)

**Review by deepesh-vasthimal-cko — CHANGES_REQUESTED** [review]
Minor changes other LGTM. Can you also please add a screen shot to PR for my benefit to understand what does the new header look like. Thanks.

File: Source/UI/NewUI/PaymentForm/View/PaymentHeaderView.swift (hunk 6 above)
Comment (deepesh-vasthimal-cko) [reply Source/UI/NewUI/PaymentForm/View/PaymentHeaderView.swift:103]: Thanks, Something to add in backlog for future improvement if needed.

**Comment by deepesh-vasthimal-cko** [comment]
Something wrong with the builds : Error: Both API Token and WebhookURL are empty. You need to provide one of them. If you want to use incoming webhooks provide the webhook url. If you want to use a bot to send a message provide the bot API token

File: Source/UI/NewUI/PaymentForm/ViewController/PaymentViewController.swift (hunk 7 above)
Comment (ehab-al-cko) [reply Source/UI/NewUI/PaymentForm/ViewController/PaymentViewController.swift:326]: It can be tested using UI test

**Review by deepesh-vasthimal-cko — APPROVED** [review]

**Pull request merged**